File: projects/engineering.purdue.edu/programmingwithobjects/pwocode/chap10/ExceptionUsage4.cc
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object-Oriented Programming with C++ and Java
//
// Book Author:   Avinash Kak
//
// Chapter:     Chapter 10  Handling Exceptions
//
// Section:     Section 10.3  Some Usage Patterns For Exception Handling In C++
//



//ExceptionUsage4.cc

#include <iostream>
using namespace std;

class MyException {};
class Err {};

void f( int j ) throw( MyException, Err ) {                       //(A) 
    if ( j == 1 ) throw MyException(); 
    if ( j == 2 ) throw Err(); 
}

int main()
{
    try {
        f( 1 );
    } catch( MyException ) { 
        cout << "caught MyException -- arg must be 1" << endl; 
    } catch( Err ) { 
        cout << "caught Err -- arg must be 2" << endl;
    }

    try {  
        f( 2 );
    } catch( MyException ) { 
        cout << "caught MyException -- arg must be 1" << endl; 
    } catch( Err ) { 
        cout << "caught Err -- arg must be 2" << endl; 
    }

    return 0;
}

File: projects/demos/engineering.purdue.edu/programmingwithobjects/pwocode/chap10/ExceptionUsage4.java
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object-Oriented Programming with C++ and Java
//
// Book Author:  Avinash Kak
//
// Chapter:     Chapter 10  Handling Exceptions
//
// Section:     Section 10.6  Some Usage Patterns For Exception Handling In Java
//



//ExceptionUsage4.java

class MyException extends Exception {}
class Err extends Exception {}

class Test {
    static void f() throws MyException {                          //(A)
        throw new MyException();
    }
    static void g() throws Err {                                  //(B)
        throw new Err();
    }
    static void h() throws MyException, Err {                     //(C)
        f();
        g();
    }

    public static void main( String[] args ) {
        try {
            h();                                                  //(D)
        } catch( MyException e ) { 
            System.out.println( "caught MyException" );
        } catch( Err e ) { 
            System.out.println( "caught Err" );
        }
    }
}